src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/Formatting/FormattingContext.cs

```diff
@@ -194,25 +196,37 @@ public async Task<FormattingContext> WithTextAsync(SourceText changedText)
             var engine = OriginalSnapshot.Project.GetProjectEngine();
             var importSources = new List<RazorSourceDocument>();
 
-            if (OriginalSnapshot is DefaultDocumentSnapshot documentSnapshot)
+            var imports = OriginalSnapshot.GetImports();
+            foreach (var import in imports)
             {
-                var imports = documentSnapshot.State.GetImports((DefaultProjectSnapshot)OriginalSnapshot.Project);
-                foreach (var import in imports)
-                {
-                    var sourceText = await import.GetTextAsync();
-                    var source = sourceText.GetRazorSourceDocument(import.FilePath, import.TargetPath);
-                    importSources.Add(source);
-                }
+                var sourceText = await import.GetTextAsync();
+                var source = sourceText.GetRazorSourceDocument(import.FilePath, import.TargetPath);
+                importSources.Add(source);
             }
 
             var changedSourceDocument = changedText.GetRazorSourceDocument(OriginalSnapshot.FilePath, OriginalSnapshot.TargetPath);
 
             var codeDocument = engine.ProcessDesignTime(changedSourceDocument, OriginalSnapshot.FileKind, importSources, OriginalSnapshot.Project.TagHelpers);
 
+            ValidateComponents(CodeDocument, codeDocument);
+
             var newContext = Create(Uri, OriginalSnapshot, codeDocument, Options, Range, IsFormatOnType);
             return newContext;
         }
 
+        /// <summary>
+        /// It can be difficult in the testing infrastructure to correct constructs input files that work consistently across
+        /// context changes, so this method validates that the number of components isn't changing due to lost tag help info.
+        /// Without this guarantee its hard to reason about test behaviour/failures.
+        /// </summary>
+        [Conditional("DEBUG")]
```

Comment:
Any reason to not just always have this on? I imagine this could be quite helpful just to always verify and wouldn't hurt much

Reply:
If this issue had ever come up at runtime we'd know about it other ways (eg, classification would break and elements would change color), but mainly I didn't want to for perf reasons. This descends through every node in a razor document, twice, and does so somewhere between 5 and 10 times per format operation.

Not that formatting is quick, but no point making it worse :)

Reply:
lol alright fair enough. Happy to have this though!